src/main/java/de/tum/cit/aet/artemis/programming/service/localci/scaparser/strategy/sarif/SarifParser.java

```diff
@@ -0,0 +1,142 @@
+import de.tum.cit.aet.artemis.programming.service.localci.scaparser.format.sarif.ReportingDescriptorReference;
+import de.tum.cit.aet.artemis.programming.service.localci.scaparser.format.sarif.Result;
+import de.tum.cit.aet.artemis.programming.service.localci.scaparser.format.sarif.Run;
+import de.tum.cit.aet.artemis.programming.service.localci.scaparser.format.sarif.SarifLog;
+import de.tum.cit.aet.artemis.programming.service.localci.scaparser.format.sarif.ToolComponent;
+import de.tum.cit.aet.artemis.programming.service.localci.scaparser.strategy.ParserStrategy;
+
+/**
+ * Implements parts of the SARIF OASIS standard version 2.1.0.
+ *
+ * @see <a href="https://docs.oasis-open.org/sarif/sarif/v2.1.0/errata01/os/sarif-v2.1.0-errata01-os-complete.html">SARIF specification</a>
+ */
+public class SarifParser implements ParserStrategy {
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
+
+    private final StaticCodeAnalysisTool tool;
+
+    private final RuleCategorizer ruleCategorizer;
+
+    public SarifParser(StaticCodeAnalysisTool tool, RuleCategorizer ruleCategorizer) {
+        this.tool = tool;
+        this.ruleCategorizer = ruleCategorizer;
+    }
+
+    @Override
+    public StaticCodeAnalysisReportDTO parse(String reportContent) {
+        SarifLog sarifLog;
+        try {
+            sarifLog = objectMapper.readValue(reportContent, SarifLog.class);
+        }
+        catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+
+        Run run = sarifLog.getRuns().getFirst();
+        ToolComponent driver = run.getTool().getDriver();
+
+        List<ReportingDescriptor> rules = driver.getRules().orElse(List.of());
+
+        // Rule ids are not guaranteed to be unique. Use the first occurring for rule lookup.
+        Map<String, ReportingDescriptor> ruleOfId = rules.stream().collect(Collectors.toMap(ReportingDescriptor::getId, Function.identity(), (first, next) -> first));
+
+        List<Result> results = run.getResults().orElse(List.of());
+        List<StaticCodeAnalysisIssue> issues = results.stream().map(result -> processResult(result, driver, ruleOfId)).toList();
+
+        return new StaticCodeAnalysisReportDTO(tool, issues);
+    }
+
+    private StaticCodeAnalysisIssue processResult(Result result, ToolComponent driver, Map<String, ReportingDescriptor> ruleOfId) {
+        PhysicalLocation location = result.getLocations().flatMap(locations -> locations.stream().findFirst()).flatMap(Location::getPhysicalLocation)
+                .orElseThrow(() -> new RuntimeException("Location needed"));
+
+        URI uri = URI.create(location.getArtifactLocation().flatMap(ArtifactLocation::getUri).orElseThrow(() -> new RuntimeException("File path needed")));
+        String path = uri.getPath();
+
+        Region region = location.getRegion().orElseThrow(() -> new RuntimeException("Region must be present"));
```

Comment:
Would it be better to emit a warning and skip this issue instead? Then the other issues could still be reported to the user.

Alternatively, some tools might not emit a location in case the whole file is affected. Could we therefore assume line 0 column 0 as location if it is not present?

Reply:
The standard does require a location and a region to be present if the result is applicable to a file. The region can cover the entire file or just point at the start of the file if needed.
But it can happen, that an issue is applicable to no file. Then no location would be given. In this case we also wouldn't be able to report a filename.

If we want to support results with missing locations we would need to make sure that the rest of the application can handle missing values.